### DISCUSSION TITLE
`cargo nextest run` is much faster

### DISCUSSION THREAD

**Discussion by RCasatta** [rust-bitcoin/rust-bitcoin#1827] [General]
Just wanted to share that once tests are build :

- `time cargo test` takes about 7 seconds to run, because tests are run in parallel inside the same binary, but in serial for each binary.
- `cargo nextest run` runs binaries in parallel and finish in just 1 second.

Obviously `nextest` requires to install other stuff and if the CI don't cache it, it will result in a slower global run, but locally or in nix environment may be convenient

**Comment by Kixunil** [comment]
Interesting, I think we could cache it. 1 second is awesome time but I guess it requires a fair bit of parallelism to achieve in practice?

**Reply by RCasatta** [reply]
I have an 8 cores/16 threads machine...

**Reply by sanket1729** [reply]
Curious, what does the threads mean in this context? I have only heard about cores per machine and execution units per core.

**Reply by dpc** [reply]
https://en.wikipedia.org/wiki/Hyper-threading

**Reply by sanket1729** [reply]
Thank you

**Reply by tcharding** [reply]
Hyperthreading enables some side channel attacks. I heard a kernel hacker talking about it after spectre and the kernel cannot mitigate all the risks so he advised the security conscious to disable hyperthreading, you loose up to 30% performance. If I understand correctly, on your local machine, the risk is that buggy web code will attack the other processes you have running on your machine.

- https://docs.kernel.org/admin-guide/hw-vuln/spectre.html

**Reply by tcharding** [reply]
@apoelstra I thought of this this morning when you were talking about your build machine, so coincidental that it came up again on the same day.

**Comment by stevenroose** [comment]
Unfortunately, especially in CI, the build takes an order of magnitude longer than running the tests, right?

**Reply by dpc** [reply]
I just peeked https://github.com/rust-bitcoin/rust-bitcoin/actions/runs/4888514794/jobs/8726236976 and it seems like no caching is being used whatsoever? That's a lot of time wasted, unless I'm missing something here.

In Fedimint (which includes rust-bitcoin and tons of other crates (unfortunately), we get build of the whole workspace under 4 minutes, and most (long and slow) integration tests under another 5 minutes. A combination of parallelism, using buildjet, and efficient caching using Nix.

Even without using Nix & cachix and buildjet there's lots that can be done to bring these down: https://www.reillywood.com/blog/rust-faster-ci/
